### PR TITLE
Pre consensus fix

### PR DESCRIPTION
**Pull request by alonmuroch**
- fixes round change (not prev prepared) pre-consensus justification
- fixing adding pre-consensus justification when calling baseRunner.decide

### CODE REVIEW

File: ssv/runner.go [hunk 1]
```diff
@@ -48,6 +48,7 @@ type BaseRunner struct {
 	BeaconRoleType types.BeaconRole
 
 	// highestDecidedSlot holds the highest decided duty slot and gets updated after each decided is reached
+	// used to validate pre-consensus justification which can not have any lower duty slot than highestDecidedSlot
 	highestDecidedSlot spec.Slot
```
Comment (GalRogozinski) [inline ssv/runner.go:52]: now that the controller's height is also the highest decided slot are we sure we really need this?

**Review by GalRogozinski — COMMENTED** [review]
I don't see any changes to Baserunner.Decide() like the description said.

File: ssv/spectest/tests/msgcontainer/duplicate_msg.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+// DuplicateMsg tests adding a duplicate msg (same signer)
+func DuplicateMsg() tests.SpecTest {
+	ks := testingutils.Testing4SharesSet()
+
+	return &SpecTest{
+		Name: "duplicate msg",
+		MsgsToAdd: []*types.SignedPartialSignatureMessage{
+			testingutils.PreConsensusRandaoDifferentSignerMsgV(ks.Shares[1], ks.Shares[1], 1, 1, spec.DataVersionBellatrix),
```
Comment (GalRogozinski) [inline ssv/spectest/tests/msgcontainer/duplicate_msg.go:17]: why not use `PreConsensusRandaoMsgV`?

File: ssv/runner.go [hunk 3]
```diff
@@ -156,13 +156,20 @@ func (b *BaseRunner) basePostConsensusMsgProcessing(runner Runner, signedMsg *ty
...
-// basePartialSigMsgProcessing adds an already validated partial msg to the container, checks for quorum and returns true (and roots) if quorum exists
+// basePartialSigMsgProcessing adds an already validated partial msg to the container,
+// checks for quorum and returns true (and roots) if we should process the msg
 func (b *BaseRunner) basePartialSigMsgProcessing(
 	signedMsg *types.SignedPartialSignatureMessage,
 	container PartialSignatureContainer,
 ) (bool, [][32]byte, error) {
-	container[signedMsg.Signer] = signedMsg
 
+	if b.Share.HasQuorum(len(container)) {
```
Comment (alonmuroch) [inline ssv/runner.go:166]: need some explanation here

Comment (GalRogozinski) [reply ssv/runner.go:163]: check that quorum happened once

File: ssv/pre_consensus_justification.go [hunk 4]
```diff
@@ -9,23 +9,15 @@ import (
 
 // correctQBFTState returns true if QBFT controller state requires pre-consensus justification
 func (b *BaseRunner) correctQBFTState(msg *qbft.SignedMessage) bool {
```
Comment (GalRogozinski) [inline ssv/pre_consensus_justification.go:11]: func name doesn't reflect functionality. 
requiresPreCosensus

File: ssv/pre_consensus_justification.go [hunk 5]
```diff
@@ -9,23 +9,15 @@ import (
...
-
-	return firstHeightNotDecided || notFirstHeightDecided
+	// mind race conditions
+	return b.QBFTController.Height < msg.Message.Height
 }
 
-// shouldProcessingJustificationsForHeight returns true if pre-consensus justification should be processed, false otherwise
-func (b *BaseRunner) shouldProcessingJustificationsForHeight(msg *qbft.SignedMessage) bool {
+// shouldProcessingJustifications returns true if pre-consensus justification should be processed, false otherwise
+func (b *BaseRunner) shouldProcessingJustifications(msg *qbft.SignedMessage) bool {
```
Comment (GalRogozinski) [inline ssv/pre_consensus_justification.go:17]: rename shouldProcessJustification

File: ssv/pre_consensus_justification.go [hunk 6]
```diff
@@ -9,23 +9,15 @@ import (
...
-
-	// firstHeightNotDecided is true if height == 0 (special case) and did not start yet
-	firstHeightNotDecided := inst == nil && b.QBFTController.Height == msg.Message.Height && msg.Message.Height == qbft.FirstHeight
-
-	// notFirstHeightDecided returns true if height != 0, height decided and the message is for next height
-	notFirstHeightDecided := decidedInstance && msg.Message.Height > qbft.FirstHeight && b.QBFTController.Height+1 == msg.Message.Height
-
-	return firstHeightNotDecided || notFirstHeightDecided
+	// mind race conditions
+	return b.QBFTController.Height < msg.Message.Height
```
Comment (GalRogozinski) [inline ssv/pre_consensus_justification.go:13]: 1. no need for a separate func, just add it to shouldProcessJustifications
2. this means after the first valid pre-consensus justification msg, the next msg will not process pre-consensus justification. This means we need dedicated tests for it

File: ssv/runner.go (hunk 1 above)
Comment (GalRogozinski) [reply ssv/runner.go:52]: we can remove it and use the controller's height instead (please add tests)

**Review by MatheusFranco99 — APPROVED** [review]

File: ssv/runner.go [hunk 7]
```diff
@@ -57,23 +54,16 @@ func NewBaseRunner(
 	controller *qbft.Controller,
 	beaconNetwork types.BeaconNetwork,
 	beaconRoleType types.BeaconRole,
-	highestDecidedSlot spec.Slot,
```
Comment (alonmuroch) [inline ssv/runner.go:60]: Explanation: Since we switched to having qbft instances start with duty slot as height there is no need for highest decided anymore

File: ssv/runner.go (hunk 3 above)
Comment (alonmuroch) [reply ssv/runner.go:149]: but why does it need to change?

File: ssv/runner_state.go [hunk 8]
```diff
@@ -12,8 +12,8 @@ import (
 
 // State holds all the relevant progress the duty execution progress
 type State struct {
-	PreConsensusContainer  *PartialSigContainer
-	PostConsensusContainer *PartialSigContainer
+	PreConsensusContainer  PartialSignatureContainer
```
Comment (alonmuroch) [inline ssv/runner_state.go:15]: I don't fully understand this change? why is this part of this PR?

File: ssv/partial_sig_container.go [hunk 9]
```diff
@@ -1,39 +1,25 @@
 package ssv
 
 import (
-	"encoding/hex"
+	"bytes"
```
Comment (alonmuroch) [inline ssv/partial_sig_container.go:4]: why are all those changes (partial sig container) relevant to this PR?

File: ssv/pre_consensus_justification.go [hunk 10]
```diff
@@ -1,35 +1,23 @@
...
-	firstHeightNotDecided := inst == nil && b.QBFTController.Height == msg.Message.Height && msg.Message.Height == qbft.FirstHeight
-
-	// notFirstHeightDecided returns true if height != 0, height decided and the message is for next height
-	notFirstHeightDecided := decidedInstance && msg.Message.Height > qbft.FirstHeight && b.QBFTController.Height+1 == msg.Message.Height
-
-	return firstHeightNotDecided || notFirstHeightDecided
-}
-
-// shouldProcessingJustificationsForHeight returns true if pre-consensus justification should be processed, false otherwise
-func (b *BaseRunner) shouldProcessingJustificationsForHeight(msg *qbft.SignedMessage) bool {
```
Comment (alonmuroch) [inline ssv/pre_consensus_justification.go:25]: why  did we remove all those checks? not all message carry pre-consensus justification

File: ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go [hunk 11]
```diff
@@ -28,9 +28,8 @@ func ProposeBlindedBlockDecidedRegular() tests.SpecTest {
 					[]types.OperatorID{1, 2, 3},
 					qbft.Height(testingutils.TestingDutySlotV(spec.DataVersionBellatrix)),
 					testingutils.ProposerMsgID,
-					testingutils.TestProposerConsensusDataBytsV(spec.DataVersionBellatrix),
-				),
-				nil),
+					testingutils.TestProposerConsensusDataBytsV(ks, spec.DataVersionBellatrix),
```
Comment (alonmuroch) [inline ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go:31]: we have a ton of those changes which just add "changed" files to the PR

File: ssv/runner_state.go (hunk 8 above)
Comment (GalRogozinski) [reply ssv/runner_state.go:15]: See commit 91ba801c

You are its author. So maybe you can answer the question 😅

File: ssv/partial_sig_container.go (hunk 9 above)
Comment (GalRogozinski) [reply ssv/partial_sig_container.go:4]: https://github.com/bloxapp/ssv-spec/commit/91ba801c3fd2c1734800eefa980e10e2055621b9

You added it here but I agree. Were you working on something else, or you want to keep it here?

File: ssv/pre_consensus_justification.go (hunk 10 above)
Comment (GalRogozinski) [reply ssv/pre_consensus_justification.go:25]: We now have simplified this. Compare more closely with `ShouldProcessJustifications`

File: ssv/spectest/generate/state_comparison/newduty_MultiStartNewRunnerDutySpecTest/duplicate_duty_finished/attester.json [hunk 12]
```diff
@@ -147,7 +141,7 @@
 								"CommitteeLength": 128,
 								"CommitteesAtSlot": 36,
 								"ValidatorCommitteeIndex": 11,
-								"ValidatorSyncCommitteeIndices": null
+								"ValidatorSyncCommitteeIndices": []
```
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/newduty_MultiStartNewRunnerDutySpecTest/duplicate_duty_finished/attester.json:144]: Looks like there was a problem with passing tests.
Is it important? I remember once I asked you about something similar and you said that changes between null or empty arrays are not so important.

File: ssv/runner.go [hunk 13]
```diff
@@ -234,6 +212,11 @@ func (b *BaseRunner) hasRunningDuty() bool {
 }
 
 func (b *BaseRunner) ShouldProcessDuty(duty *types.Duty) error {
+	// first height special case
```
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply ssv/runner.go:215]: *Why do we need it?*

It is a special case because of an implementation detail. Currently we bootstrap the Controller's state with 0. Also the first duty can also be decided with height 0. In all other cases the duty will be decided on a height greater than the controller.

We can do a smart fix in spec so that 0 height will work like any other height and won't need special conditions in our algos (not in this PR).

However, even if we do that maybe we should have first height tests. We can't do check for all heights but first height is suspect for bugs.

*do we test it?*
See `valid_first_height.go`

File: ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go (hunk 11 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go:31]: Many of them come from your commit
0266dd81

I'll try to clean it up

File: ssv/runner_state.go (hunk 8 above)
Comment (GalRogozinski) [reply ssv/runner_state.go:15]: The previous PartialSigMessageContainer can't return the full PartialSigMessage which is a requirement for putting the preconsensus justification in the CD.
I initially thought you did this as part of the spec-rethink but now I think you did this for this PR and I propose we leave those changes

File: ssv/partial_sig_container.go (hunk 9 above)
Comment (GalRogozinski) [reply ssv/partial_sig_container.go:4]: The previous PartialSigMessageContainer can't return the full PartialSigMessage which is a requirement for putting the preconsensus justification in the CD.
I initially thought you did this as part of the spec-rethink but now I think you did this for this PR and I propose we leave those changes

**Review by MatheusFranco99 — APPROVED** [review]
Looks good to me

File: ssv/runner.go (hunk 3 above)
Comment (alonmuroch) [reply ssv/runner.go:149]: ?

**Comment by alonmuroch** [comment]
approved

**Review by GalRogozinski — DISMISSED** [review]

File: ssv/runner.go [hunk 14]
```diff
@@ -155,30 +139,24 @@ func (b *BaseRunner) basePostConsensusMsgProcessing(runner Runner, signedMsg *ty
...
+		container[signedMsg.Signer] = signedMsg
+		// return false if we already have quorum
+		return false, container.Roots(), nil
 	}
 
-	return anyQuorum, roots, nil
+	container[signedMsg.Signer] = signedMsg
+	if b.Share.HasQuorum(len(container)) {
+		return true, container.Roots(), nil
+	}
```
Comment (moshe-blox) [inline ssv/runner.go:158]: Isn't this skipping the check that the signers have signed the same Beacon roots?

I imagine that if operators sign different aggregation messages, and the signing roots aren't checked, then the Runner may falsely think it has quorum and proceed to fail reconstructing the signature. If true, this means that one faulty operator may ruin the duty, no?

On another note, it seems to me that `container.Roots()` may cause another consensus issue. Since it only returns the signing roots from one `SignedPartialSignatureMessage`, this means that if the operators have signed different roots, then we'd only try to reconstruct & submit those from the first message in the container (which is a map so the order is non-deterministic).

In case anyone wants to know why multiple signing roots are necessary:
in sync committee aggregation duties, although extremely rare, one validator may participate in several subcommittees, in which case there will be several roots under `signedMsg.Message.Messages`, one for each subcommittee.

Comment (GalRogozinski) [reply ssv/runner.go:158]: So double check me please @moshe-blox:

1. Note that before `basePartialSigMsgProcessing` is called we always call some validation method. See:
https://github.com/bloxapp/ssv-spec/blob/daec5450ce7d808a7e643081a5109b4eb7c55f07/ssv/pre_consensus_justification.go#L20
https://github.com/bloxapp/ssv-spec/blob/0a414236cef5eca4dc0f7da464034325869a6477/ssv/runner_validations.go#L29
https://github.com/bloxapp/ssv-spec/blob/0a414236cef5eca4dc0f7da464034325869a6477/ssv/runner_validations.go#L12

Those validations should take care of the problem you mentioned. I will add method comments though to make this clear though.

2. Having said this it might be smart when it comes to the implementation to add this check as part of a defensive programming strategy. If a bug is introduced in validation you have a fall back mechanism. Or you can add comments and have good tests (your choice, just giving my 2 cents.

3. Same with the problem concerning `container.Roots()`. Not sure if I need to add comments though because this is just a getter method to what was inserted to the container.

Comment (GalRogozinski) [reply ssv/runner.go:158]: actually @moshe-blox I think the existing comment is fine:
https://github.com/bloxapp/ssv-spec/blob/ecc6a3f944d3dcb2387be71eb29920276f11a131/ssv/runner.go#L142

If you think it should be clearer you can make a suggestion :-)

Comment (moshe-blox) [reply ssv/runner.go:158]: hey @GalRogozinski 

Thanks for the clarifying, I see that this issue can't happen because the roots must be identical to what the verifier got from their own Beacon node.

File: ssv/partial_sig_container.go [hunk 15]
```diff
@@ -43,10 +29,43 @@ func (ps *PartialSigContainer) ReconstructSignature(root [32]byte, validatorPubK
 	return signature.Serialize(), nil
 }
 
-func (ps *PartialSigContainer) HasQuorum(root [32]byte) bool {
-	return uint64(len(ps.Signatures[rootHex(root)])) >= ps.Quorum
+// SignatureForRoot returns a map of signer and signature for a specific root
+func (ps PartialSignatureContainer) SignatureForRoot(root [32]byte) map[types.OperatorID][]byte {
```
Comment (y0sher) [inline ssv/partial_sig_container.go:33]: not trying to nit pick but I think this function should be named 
```suggestion
func (ps PartialSignatureContainer) SignaturesForRoot(root [32]byte) map[types.OperatorID][]byte {
```

File: ssv/partial_sig_container.go [hunk 16]
```diff
@@ -43,10 +29,43 @@ func (ps *PartialSigContainer) ReconstructSignature(root [32]byte, validatorPubK
...
+	if len(ps) > 0 {
+		ret := make([][32]byte, 0)
+		for _, sigMsg := range ps {
+			for _, msg := range sigMsg.Message.Messages {
+				ret = append(ret, msg.SigningRoot)
+			}
+			break // only need to iterate first msg
+		}
+		return ret
+	}
```
Comment (y0sher) [inline ssv/partial_sig_container.go:56]: we could do 
```golang
if len(ps) =< 0 {
return nil // or [][32]byte{}
}
```
 to reduce nesting and add readability

File: ssv/pre_consensus_justification.go [hunk 17]
```diff
@@ -39,8 +27,8 @@ func (b *BaseRunner) validatePreConsensusJustifications(data *types.ConsensusDat
 		return errors.New("wrong beacon role")
 	}
 
-	if data.Duty.Slot <= highestDecidedDutySlot {
-		return errors.New("duty.slot <= highest decided slot")
+	if qbft.Height(data.Duty.Slot) <= b.QBFTController.Height && b.QBFTController.Height != qbft.FirstHeight {
```
Comment (y0sher) [inline ssv/pre_consensus_justification.go:30]: if we are on first height (which will never happen I guess) then its ok that the duty height is smaller than the controller height? I understand why its ok that its equal but not lower.

File: ssv/runner.go [hunk 18]
```diff
@@ -167,30 +151,24 @@ func (b *BaseRunner) basePostConsensusMsgProcessing(runner Runner, signedMsg *ty
...
-			continue
-		}
-
-		quorum := container.HasQuorum(msg.SigningRoot)
-		if quorum {
-			roots = append(roots, msg.SigningRoot)
-			anyQuorum = true
-		}
+
+	if b.Share.HasQuorum(len(container)) {
```
Comment (y0sher) [inline ssv/runner.go:161]: it seems like `HasQuorum` just checks the number of sigs we have, don't we want to also check validity before we say we have quorum ?

File: ssv/spectest/generate/state_comparison/tests_MultiMsgProcessingSpecTest/consensus_decided_invalid_value/aggregator.json [hunk 19]
```diff
@@ -2,19 +2,152 @@
...
+										"Signer": 2
+								},
+								"3": {
+										"Message": {
+												"Type": 2,
+												"Slot": 12,
+												"Messages": [
+														{
+																"PartialSignature": "qFQDMkcUT9KZK3wO4GHZKlEZnJKWbhULlEMkiYC0igtS+snZxJ5tfnZ5oxKvIPRcBVMI0gP/rhHuEgGzpwBuVfuFYclsirb3DZVgPJcdVBjSkhkAt4polkJ/BKEHBdC5",
+																"SigningRoot": [
```
Comment (y0sher) [inline ssv/spectest/generate/state_comparison/tests_MultiMsgProcessingSpecTest/consensus_decided_invalid_value/aggregator.json:108]: can we make this hex encoded like other roots/signatures/pubkeys?

File: ssv/spectest/tests/msgcontainer/quorum_post_consensus_msg.go [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+// QuorumPostConsensusMsg tests adding a quorum of post consensus message to container
+func QuorumPostConsensusMsg() tests.SpecTest {
+	ks := testingutils.Testing4SharesSet()
+
+	return &SpecTest{
+		Name: "quorum post consensus message",
+		MsgsToAdd: []*types.SignedPartialSignatureMessage{
+			testingutils.PostConsensusSyncCommitteeContributionMsg(ks.Shares[1], 1, ks),
+			testingutils.PostConsensusSyncCommitteeContributionMsg(ks.Shares[2], 2, ks),
+			testingutils.PostConsensusSyncCommitteeContributionMsg(ks.Shares[3], 3, ks),
```
Comment (y0sher) [inline ssv/spectest/tests/msgcontainer/quorum_post_consensus_msg.go:18]: should we test with messages other than synccommitteecontrib?

File: ssv/spectest/tests/msgcontainer/test.go [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+		for i, m := range msg.Message.Messages {
+			sig, lastErr = c.ReconstructSignature(m.SigningRoot, testingutils.TestingValidatorPubKey[:])
+			require.EqualValues(t, test.PostReconstructedSignature[i], hex.EncodeToString(sig))
+		}
+		if len(test.ExpectedErr) == 0 {
+			require.NoError(t, lastErr)
+		} else {
+			require.EqualError(t, lastErr, test.ExpectedErr)
+		}
+	}
```
Comment (y0sher) [inline ssv/spectest/tests/msgcontainer/test.go:55]: what about checking ExpectedErr even if postreconstructedsignature is empty? or there's no such test?

File: ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go [hunk 22]
```diff
@@ -19,8 +19,8 @@ func ValidDecided10Operators() tests.SpecTest {
 				Name:                    "sync committee contribution",
 				Runner:                  testingutils.SyncCommitteeContributionRunner(ks),
 				Duty:                    &testingutils.TestingSyncCommitteeContributionDuty,
-				Messages:                testingutils.SSVDecidingMsgsV(testingutils.TestSyncCommitteeContributionConsensusData, ks, types.BNRoleSyncCommitteeContribution),
-				PostDutyRunnerStateRoot: "a991b6470a8c7a55f4ce89aea91925c2d80a9a8c4258545cc2fb17cabc388719",
+				Messages:                testingutils.SSVDecidingMsgsV(testingutils.TestSyncCommitteeContributionConsensusData(ks), ks, types.BNRoleSyncCommitteeContribution),
+				PostDutyRunnerStateRoot: "ec47a2194bab81891bd581eaf427af211b53e13d1f77a6703eeb56a693caa878",
```
Comment (y0sher) [inline ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go:23]: this is changed because preconsensus now affects the duty runner state?

File: ssv/spectest/tests/runner/full_happy_flow.go [hunk 23]
```diff
@@ -69,19 +69,12 @@ func FullHappyFlow() tests.SpecTest {
...
-							testingutils.SSVMsgAggregator(nil, testingutils.PostConsensusAggregatorMsg(ks.Shares[2], 2)),
-							testingutils.SSVMsgAggregator(nil, testingutils.PostConsensusAggregatorMsg(ks.Shares[3], 3)),
-						}...,
-					)...,
+					testingutils.SSVDecidingMsgsV(testingutils.TestAggregatorConsensusData(ks), ks, types.BNRoleAggregator), // consensus
+					[]*types.SSVMessage{ // post consensus
+						testingutils.SSVMsgAggregator(nil, testingutils.PostConsensusAggregatorMsg(ks.Shares[1], 1)),
+						testingutils.SSVMsgAggregator(nil, testingutils.PostConsensusAggregatorMsg(ks.Shares[2], 2)),
+						testingutils.SSVMsgAggregator(nil, testingutils.PostConsensusAggregatorMsg(ks.Shares[3], 3)),
+					}...,
```
Comment (y0sher) [inline ssv/spectest/tests/runner/full_happy_flow.go:77]: this duty now doesn't have pre consensus msgs?

File: ssv/spectest/tests/runner/pre_consensus_justifications/inconsisten_root_count.go
Comment (y0sher) [inline ssv/spectest/tests/runner/pre_consensus_justifications/inconsisten_root_count.go:1]: not from this pr but typo in file name

**Review by y0sher — CHANGES_REQUESTED** [review]
Big PR. see my comments, mostly clarifications.

File: ssv/pre_consensus_justification.go (hunk 17 above)
Comment (GalRogozinski) [reply ssv/pre_consensus_justification.go:30]: If it is small or equal it returns an error.
On first height it can only be equal.

Comment (GalRogozinski) [reply ssv/pre_consensus_justification.go:30]: you mean if firstHeight would be a number different than 0?
If it will be 1 then we won't need this special condition in all those places in the code

File: ssv/partial_sig_container.go (hunk 15 above)
Comment (GalRogozinski) [reply ssv/partial_sig_container.go:33]: done

File: ssv/partial_sig_container.go (hunk 16 above)
Comment (GalRogozinski) [reply ssv/partial_sig_container.go:56]: done

File: ssv/runner.go (hunk 18 above)
Comment (GalRogozinski) [reply ssv/runner.go:161]: for a message to enter the container it should pass validity checks

File: ssv/spectest/generate/state_comparison/tests_MultiMsgProcessingSpecTest/consensus_decided_invalid_value/aggregator.json (hunk 19 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/tests_MultiMsgProcessingSpecTest/consensus_decided_invalid_value/aggregator.json:108]: not for this PR

File: ssv/spectest/tests/msgcontainer/quorum_post_consensus_msg.go (hunk 20 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/msgcontainer/quorum_post_consensus_msg.go:18]: maybe...
when I look at this now not sure why this test is part of the PR (it is something Alon added)...
The PR should concentrate on preconsensus...

nevertheless I am adding you point in #264

File: ssv/spectest/tests/msgcontainer/test.go (hunk 21 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/msgcontainer/test.go:55]: Currently there is a trick in tests that Alon did to have a non empty `PostReconstructedSignature` filled with empty strings.

In another test he kept the `PostReconstructedSignature` empty to not have an error...
So I'd rather not change his design at this point

File: ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go (hunk 22 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go:23]: the runner contains all msg containers that includes messages with preconsensus justifications, yes...
But I don't see in the logic where we add prec justification for this test... will look into it...

File: ssv/spectest/tests/runner/full_happy_flow.go (hunk 23 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/full_happy_flow.go:77]: `testingutils.SSVDecidingMsgsV(testingutils.TestSyncCommitteeConsensusData, ks, types.BNRoleSyncCommittee), // consensus
`

this line adds precon messages
see `pre_consensus_full_happy_flow`

File: ssv/spectest/tests/runner/pre_consensus_justifications/inconsisten_root_count.go
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/pre_consensus_justifications/inconsisten_root_count.go:1]: on one hand the fix for this doesn't belong in this PR...
on the other hand getting 3 busy people to approve a typo fix on a dedicated PR will take ages...

File: ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go (hunk 22 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go:23]: the commit 66c1585f has this [change](https://github.com/bloxapp/ssv-spec/commit/66c1585ff0a1ef79b31ccd2453a37314cb5a0ea8#diff-28956438e46e71f8b726f9fe106f47ac7cfca29ae303a20aacc6e7837a29b506R21)

This should be overridden by deneb imo and the root may need to change again...
I will resolve conflicts and then regenrate jsons and see what's going on

File: ssv/spectest/tests/runner/pre_consensus_justifications/inconsisten_root_count.go
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/pre_consensus_justifications/inconsisten_root_count.go:1]: but it will add more conflicts so I won't fix

File: ssv/partial_sig_container.go [hunk 24]
```diff
@@ -43,10 +29,44 @@ func (ps *PartialSigContainer) ReconstructSignature(root [32]byte, validatorPubK
...
+		break // only need to iterate first msg
+	}
+	return ret
 }
 
-func rootHex(r [32]byte) string {
-	return hex.EncodeToString(r[:])
+// AllSorted returns ordered by signer array of signed messages
+func (ps PartialSignatureContainer) AllSorted() []*types.SignedPartialSignatureMessage {
+	ret := make([]*types.SignedPartialSignatureMessage, 0)
```
Comment (moshe-blox) [inline ssv/partial_sig_container.go:64]: allocate with capacity of `len(ps)`

File: ssv/partial_sig_container.go [hunk 25]
```diff
@@ -43,10 +29,44 @@ func (ps *PartialSigContainer) ReconstructSignature(root [32]byte, validatorPubK
...
+	for _, m := range ps {
+		ret = append(ret, m)
+	}
+	sort.Slice(ret, func(i, j int) bool {
```
Comment (moshe-blox) [inline ssv/partial_sig_container.go:68]: for less memory allocations:
```suggestion
	ret := make([]*types.SignedPartialSignatureMessage, len(ps))
	for i, m := range ps {
		ret[i] = m
	}
```

File: ssv/aggregator.go
Comment (moshe-blox) [inline ssv/aggregator.go:1]: rename `quorum` to `shouldProcess` or the other way around to keep consistency

personally i like `quorum` bc its self-explanatory

**Review by moshe-blox — APPROVED** [review]

File: ssv/partial_sig_container.go (hunk 24 above)
Comment (GalRogozinski) [reply ssv/partial_sig_container.go:64]: will do because you copy it to impl even though ideally spec shouldn't care

File: ssv/partial_sig_container.go (hunk 25 above)
Comment (GalRogozinski) [reply ssv/partial_sig_container.go:68]: why delete sort?

File: ssv/aggregator.go
Comment (GalRogozinski) [reply ssv/aggregator.go:1]: the problem is that sometimes you have a quorum but you don't process
